### DISCUSSION TITLE
"Add to Dataset" : augment ground truth labelling with custom AI-assistant

### DISCUSSION THREAD

**Discussion by virgilemermet-alan** [langfuse/langfuse#4850] [Ideas]
### Describe the feature or potential improvement

**🎯 Improvement : make ground truth writing more efficient & user friendly**

- Current UX for ground truth writing is really basic : editing directly the JSON  ([same as langsmith](https://docs.smith.langchain.com/old/evaluation/faq/manage-datasets))
- Yet ground truth writing can be a very heavy task *(see below)*

👉 [simple] Having a “conversation “ UX would be nice _(like Pretty ✨  chat Trace)_
👉 [killer feature] include the ability to use a custom "rewriter AI assistant" when editing ground truth

### Additional information

<details>
<summary>Problem - writing ground truth is a heavy task</summary>

As an insurance company, we are building chatbots answering questions from our members. We are using different prompts depending on the kind of question. It works well, and we are now scaling our approach, which includes building  automated evaluations. For that, we are doing ground truth labelling. For ground truth labelling our bar is high in terms of : 
- **volume**: we create couple hundreds of datapoint per months
- **complexity** : our prompt are 10k+ tokens, with a dozen variable being injected (from member profiles). This makes the ground truth complex to determine for some configurations
- **quality** : our ground truth should be very carefully designed - and the task is not easy
	- we want our assistant to clarify the user's question when the answer is not deterministic
	- we sometimes want our assistant to answer **exactly** according to some predefined answers
	- we do NOT want our assistant to answer when the question is out of its scope

</details>

<details>
<summary>Our current solution : using a "rewriter assistant"</summary>

We've seen writing a ground truth is expensive : 
- Rewriting a conversation manually is tedious (_complicated questions + manually rewriting)_
- It can ONLY be performed by domain experts, not data scientist / prompt engineers
- yet it must follow guidelines regarding "expected AI behavior" (_somewhat related to prompt accessible knowledge_)

👉 Our solution has been to rely on a "rewriter AI assistant"

The principle is pretty simple : 
- LLMs are quite good at detecting their own mistake when prompted rightfully
- LLMs are good at rewriting a conversation based on a short human feedback

**Our process goes like this :** 
1. human is providing a high level feedback "chatbot should have taken into account the fact that member’s coverage is suspended to answer the question"
2. our "rewriter AI " immediately **rewrites the conversation**, using :
    - chatbot prompt + conversation
    - high level feedback (just filled by human)
    - guidelines ("how to rewrite a conversation given feedback")
3. human is proofreading the "rewritten conversation", with the ability to edit in the conversational UX
4. *Ideally, human can call the rewriter many times if initial feedback was not enough*

👉 This speeds up the ground truth writing by at least x2, while improving the quality
</details>

### Questions 

👉 Does anyone share the same problem with ground truth labelling being heavy ? Do you actually edit directly in the JSON ?
👉 What do you think about our solution/has anyone resorted to similar approaches? 
👉  @marcklingen Has this kind of feature ever been considered ? [related discussion bringing ground truth in the game](https://github.com/orgs/langfuse/discussions/1008)

**Comment by marcklingen** [comment]
Thanks for sharing! I agree that this could be much easier within Langfuse. I've seen the following workflow a couple of times:

1. Add review comments to traces with feedback on how the response could have been better. Can be either done via api (own interface), on Langfuse trace view, or via annotation queue within Langfuse.
2. Use API to get all review comments: https://api.reference.langfuse.com/#get-/api/public/comments
3. External: Rewrite/correct the response via metaprompt and review comment.
4. Add new response to Langfuse dataset via [api](https://api.reference.langfuse.com/#post-/api/public/dataset-items)/sdk with a reference to the original trace (`sourceTraceId`).

**Reply by virgilemermet-alan** [reply]
@marcklingen thanks for the answer, very clear ! 🙏 
👉 The actual rewriting has to happen in a tool outside of Langfuse 🔧 
👉 Langfuse has the features for review comment + integration with the tool ✅
